Reword a little and expand best practices, (#57)
con/opfvta-replication-2023: article/results.tex

As part of this work we have contributed substantial changes to the original OPFVTA repository, based on which we formulate a number of best practice guidelines, highly relevant in the production of reexecutable research outputs.

\subsubsection{Errors should be fatal more often than not}

> By default any Python script or a compiled C or C++ program just errors out and exits immediately when running into some unexpected operation.

Reviewer: Do we really want to be specific regarding the language? I assume it's the same for Rust and others, no? Maybe just “the majority of compiled and interpreted languages ..., though Shell in particular behaves differently.”

Author: please use suggestions... what about 
```suggestion
By default any Python script or a compiled, e.g. C or C++, program just errors out and exits immediately when running into some unexpected operation.
```
? as for having an example mentioning popular language -- could bring reader "home" and thus make it easier to relate.

Reviewer: “The majority of programming languages (including e.g. Python and C) will by default exit immediately with an error when encountering unexpected operations.
Unix shell languages, by contrast, continue with execution of the next scripted command, and return non-0 exit codes only when the list of commands is exhausted or the `exit` command is called.”

Also I think it's important to make “shell” more specific. Any interpreted language can have a shell, and even some compiled languages (e.g. haskell) have shells. Also I am not sure whether it is all Unix shells which behave like this (POSIX, Bash, csh...) or just Bash, since I only ever use Bash.

Author: > The majority of programming languages (including e.g. Python and C) will by default exit immediately ...

languages do not exit IMHO since they do not run -- executable scripts or programs do.

re shell - good point. Please suggest (and I mean - use `suggestion` - 
![image](https://github.com/con/opfvta-replication-2023/assets/39889/30a4a3ba-424c-43c0-af0c-39ccd5926c4b)
) the changes next time... let me suggest one for shell below, feel welcome to suggest another improvement.

> By default any Python script or a compiled C or C++ program just errors out and exits immediately when running into some unexpected operation.

Reviewer: ```suggestion
The majority of programming languages (including e.g. Python and C) will by default exit immediately with an error when encountering unexpected operations.
```

Reviewer: ```suggestion
By default programs written in the majority of languages (including e.g. Python and C) will just error out and exits immediately when running into an unexpected operation.
```

> Shell scripts behave differently.

Author: ```suggestion
POSIX shell (and other similar or derived ones such as bash, zsh, etc.)  behave differently.
```

By default they continue with execution of the next scripted command, and only when the list of commands is exhausted or explicit `exit` command is called, it finally exits with a non-0 error code.
As a result, even if not causing some detrimental actions, an execution of the script could continue for hours before it fails, and original error message might be lost in the flood of output making it hard or impossible to localize original problem to be fixed.

> Analogously, shell scripts treat undefined variables as a variable having an empty value, and thus not causing any error at all.

Reviewer: Is this true? I'm pretty sure bash distinguishes empty from unset, is undefined different from unset?

Author: try and show me wrong (which I can well  be)

Reviewer: ```
[deco]~ ❱ if [ -v FIFI ]; then echo "a" ; fi
[deco]~ ❱ FIFI=""; if [ -v FIFI ]; then echo "a" ; fi
a
```

Author: note - `-v` is bashism:

```shell
❯ sh -c '[ -v FIFI ] && echo a'
sh: 1: [: -v: unexpected operator
```

and what you show above is that `-v` behaves as it is documented:

```
       -v varname
              True if the shell variable varname is set (has been assigned a value).
```

and not "bash distinguishes empty from unset," which would be `-z` or `-n` depending on what bool you want out of it.  So "bash distinguishes empty from unset" is an undefined statement by itself -- bash does have operands allowing to figure out all those cases.

But what matters here is not what bash can do, but how it behaves! When variable is not defined, shell doesn't care and just uses empty value for it:

```shell
❯ sh -c 'echo ">$FIFI<"; echo a'
><
a
❯ sh -c 'FIFI=;echo ">$FIFI<"; echo a'
><
a
```

**unless** you `set -u`,

```shell
❯ sh -c 'set -u; FIFI=;echo ">$FIFI<"; echo a'
><
a
❯ sh -c 'set -u; echo ">$FIFI<"; echo a'
sh: 1: FIFI: parameter not set
```

Reviewer: ok.

As a result a command in a script like \texttt{rm -rf "\${PREFIX}/"} would do its best to try to remove everything on your harddrive in case \texttt{PREFIX} variable is not defined.

> \texttt{set -u} would change the default behavior to error out and stop execution as soon as undefined variable is attempted to be used.

Reviewer: Did we actually ever encounter this issue? I am pretty sure we only encountered the non-fatal erroring issue. Also, we mention `set -u` explicitly, which might be the more theoretical contribution (in light of us not having really encountered it), and describe `set -e` only implicitly.

Author: I don't know if we did... did we? if writing was not postponed and we added direct references -- we might have known.  But I know that I did encounter this issue MANY times in other scripts.  
So, is there anything to change?

Reviewer: Yes, but I can change it after merge if you want. If you want to change it basically we should introduce `set -e` explicitly and as the main thing, and only mention `set -u` after that. I'm pretty sure we never had that issue with this article. You just put `set -eu` directly and I didn't initially even know what the `u` was for.

Author: I guess we could do that, or just point to the sample `set -eu` stating that it is just a good safer default to use.

> As a result, even if not causing some detrimental actions, an execution of the script could continue for hours before it fails, and original error message might be lost in the flood of output making it hard or impossible to localize original problem to be fixed.
> Analogously, shell scripts treat undefined variables as a variable having an empty value, and thus not causing any error at all.
> As a result a command in a script like \texttt{rm -rf "\${PREFIX}/"} would do its best to try to remove everything on your harddrive in case \texttt{PREFIX} variable is not defined.
> \texttt{set -u} would change the default behavior to error out and stop execution as soon as undefined variable is attempted to be used.

Author: Here is from the comment above (but with the last line removed) in a correct "multiline" suggestion selected mode:

```suggestion
As a result, even if not causing some detrimental actions, an execution of the script could continue for hours before it fails, and original error message might be lost in the flood of output making it hard or impossible to localize original problem to be fixed.
This can be addressed by prepending \texttt{set -e} to the respective shell script, which changes the default behavior so that it will error out and stop execution as command exits with an error code.
Additionally, shell scripts treat undefined variables as a variable having an empty value, with empty values causing any error at all.
This can lead to numerous ill-defined behaviours, including a command such as \texttt{rm -rf "\${PREFIX}/"} removing all files on the system if \texttt{PREFIX} is not defined.
This can be addressed by prepending \texttt{set -u} so that an error is raised and execution is stopped as soon as undefined variable is attempted to be used.
```

To summarize, we advocate to include \texttt{set -eu} at the top of every shell script to guarantee it to error out as soon as any command fails or an undefined variable is attempted to be used.
Read more in the section on \href{https://www.repronim.org/module-reproducible-basics/01-shell-basics/\#fail-early}{``Command line/shell: Fail early''} subsection of the ``ReproNim Reproducible Basics Module''~\cite{repronim:reprobasics}.

\subsubsection{Avoid assuming or hard-coding absolute paths to resources}
Ensuring layout compatibility in different article execution environments is contingent on executable resources being able to find code or data required for execution.

> Explicitly hardcoded in the scripts absolute paths are likely to not exist anywhere but the original execution environment, and hence scripts guaranteed to fail.

Reviewer: Guaranteed is a bit of a stretch :)

Author: only by a very tiny margin. IMHO it is ok, but we could say
```suggestion
Explicitly hardcoded in the scripts absolute paths are likely to not exist anywhere but the original execution environment, and hence scripts almost guaranteed to fail.
```
or suggest other wording

Reviewer: “can very easily fail” I think is best

Author: then
```suggestion
Explicitly hardcoded in the scripts absolute paths are likely to not exist anywhere but the original execution environment, and hence scripts can very easily fail to fail.
```
... I don't care that much.

This problem is best avoided by adhering to YODA principles~\cite{yoda} of being able to reference all needed resources (data, scripts, container images, etc.) \emph{under} the study folder.
Use of relative paths within the study scripts then improve their portability.
Paths to external resources (scratch folders, reusable resources such as atlases etc. if not present within study folder) should be allowed to be specified dynamically via command line options or via an environment variable during execution.

\subsubsection{Avoid assuming a directory context for execution}

> Per previous recommendation, resources may be linked via relative paths, these are paths resolved based on their hierarchical location with the execution base path as the reference point.

Reviewer: I don't think we need hierarchical here, could have a 0-level hierarchy, the point is that they are relative.

Author: ```suggestion
Per previous recommendation, resources may be linked via relative paths, these are paths resolved based on their  location in the filesystem hierarchy with the execution base path as the reference point.
```
does this make the point I had in mind better?  I think "location" by itself isn't hinting on the path of hierarchical organization of the filesystem

Reviewer: yes.

It is worth to separate out at least such large steps as preprocessing, individual levels of analysis (e.g. per-subject vs whole-population), and article generation.
YODA principles~\cite{yoda} again demonstrate one way how to organize digital artifacts across such stages for maximal re-usability.
An good feature is to use a workflow platform which automatically checks for the presence of results from prior stages, and, if present, proceed to the next stage without triggering re-computation of prior stages.
In this article we rely on simple GNU Make rules to provide very basic workflow across computational stages, but there exist hundreds of workflow engines and we refer reader to a recent call for FAIR workflows to discover more~\cite{fair-workflows}.

\subsubsection{Container image size should be kept small}
Due to a lack of persistency, addressing issues in container image contents requires rebuilding, which can be a time-consuming process.
The smaller the container image, the easier it is.
In particular, when using containers, it is thus advisable to \textit{not} provide data via a package manager or via manual download inside the build script.

> A suitable alternative is to assure provision of these resources outside of the container image, e.g. by bind-mounting Git (and DataLad) directories present on the host machine or even combining them all via Git submodules mechanism as promoted by YODA principles~\cite{yoda}.

Reviewer: I think “combining” here is a bit confusing. Also, is it really related to image size? I'm not sure whether YODA is related to this point.

Author: it relates to composition... please suggest alternative to "combining" to make it less confusing

Reviewer: I can't because I'm confused about what you are describing. It doesn't matter with respect to image size whether they're modules or submodules. Just that they're in a directory which can be bind-mounted. Or am I missing something?

Author: note -- containers nicely bind things together...
But we are saying: "do not include large data into containers, just bind from outside".  But binding from outside "breaks" that binding.  Here I am promoting combining with `git submodules` so it binds it all back together .
I agree that such line of thinking is not obvious from what is written though. See how to improve?

Reviewer: ```suggestion
A suitable alternative is to assure provision of these resources outside of the container image (e.g. via Git and Datalad) and bind-mounting data present on the host machine.
```

This is an improvement I think, but it doesn't actually include your point. I think you're trying to make a separate point. We already say “assure provision”. Whether the repository uses `wget ... && tar xvf ...` or `datalad install && datalad get ...` or whether these are submodules or just copies where git/datalad are only used as an explicit download command I think is irrelevant to the point of keeping images small.

Author: ```suggestion
A suitable alternative is to assure provision of these resources outside of the container image and bind-mounting data thus downloaded to the host machine.

\subsubsection{Resources  should be bundled into a DataLad superdataset}
As external resources might change or disappear, it is beneficial to use data version control system, such as git-annex and DataLad.
The git submodule mechanism allows to bundle multiple repositories together with clear versioning information, thus following the modularity principle promoted by YODA~\cite{yoda}.
Moreover, git-annex allows for multiple data sources and data integrity verification.
```

\subsubsection{Containers should fit the scope of the underlying workflow steps}
In order to not artificially extend the workload of rebuilding a container image, it is advisable to not create a bundled container image for sufficiently distinct separate steps of the workflow.
For an example, as seen in this study, the article reexecution container image should be distinct from container images required for producing a summary meta-article.

> Complimentary, and initially possibly appearing contradictory, to the aforementioned advise we recommend to avoid building separate containers for related steps, in particular if supported by the same toolkits, and rather define and use different \emph{entry points} to the same container.

Reviewer: I think you mean complementary, right?

Author: yes! hence 

```suggestion
Complementary, and initially possibly appearing contradictory, to the aforementioned advise we recommend to avoid building separate containers for related steps, in particular if supported by the same toolkits, and rather define and use different \emph{entry points} to the same container.
```

Author: wasn't supposed to be resolved -- we should accept/commit the suggestion

Such file might be vital for debugging, and thus should not be lost.
This can be avoided by making sure that the paths used for intermediary and debugging outputs are bind-mounted to real directories on the parent system, from which they can be freely inspected.

\subsubsection{Parameterize scratch directories}
Complex workflows commonly generate large amounts of scratch data — intermediary data processing steps, with no other utility than being read by subsequent steps.
If these data are written to a hard-coded path, multiple executions will lead to race conditions, compromising one or multiple execution attempts.
This can be avoided by parameterizing the path and/or setting a default value based on a unique string (e.g. generated from the timestamps).
When using containers, this should be done at the container initiation level, as the relevant path is the path on the parent system, and not the path inside the container.

\subsubsection{Dependency versions inside container environments should be frozen as soon as feasible}

> The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in an ad-hoc environment.

Reviewer: I think it's the persistence which is the relevant feature with respect to updating single components, rather than the ad-hoc-ness. I also think ad-hoc is incorrect, it means “for a particular purpose”. In a sense containers are the ad-hoc environments more so than persistent multi-use systems.

Author: ```
❯ dict "ad hoc"
1 definition found

From WordNet (r) 3.0 (2006) [wn]:

  ad hoc
      adv 1: for one specific case; "they were appointed ad hoc"
      adj 1: often improvised or impromptu; "an ad hoc committee
             meeting"
      2: for or concerned with one specific purpose; "a coordinated
         policy instead of ad hoc decisions"
```
I guess I had more of the "often improved or impromptu" meaning -- that typical environment is just improvised sequence of installation, upgrades etc steps... so I think it fits here.  

What do you mean by persistence here - that container insides do not change?   again - please suggest alternative wording to improve

Reviewer: “The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in persistent environments.”

perhaps even better:

“The need for an entire image rebuilding also means that assuring consistent functionality in view of frequent updates is more difficult inside containers than in interactively managed environments.“

Author: I think we are also missing one more aspect... so:
```suggestion
The need for an entire image rebuilding also means that assuring consistent functionality across rebuilds in view of frequent updates is more difficult inside containers than in interactively managed environments.
```
?
